--- create3.urdf.xacro  2024-03-14 13:19:46.408251086 +0000
+++ create3.urdf.xacro        2024-03-14 13:18:23.951954841 +0000
@@ -298,7 +298,7 @@

     <gazebo>
       <plugin filename="libignition-gazebo-sensors-system.so" name="ignition::gazebo::systems::Sensors">
-        <render_engine>ogre</render_engine>
+        <render_engine>ogre2</render_engine>
       </plugin>
     </gazebo>
   </xacro:if>